torchmetrics/functional/audio/pit.py

```diff
@@ -0,0 +1,209 @@
+        _ps_dict[key] = ps
+    else:
+        ps = _ps_dict[key]  # all the permutations, shape [spk_num, perm_num]
+
+    # find the metric of each permutation
+    perm_num = ps.shape[-1]
+    # shape [batch_size, spk_num, perm_num]
+    bps = ps[None, ...].expand(batch_size, spk_num, perm_num)
+    # shape [batch_size, spk_num, perm_num]
+    metric_of_ps_details = torch.gather(metric_mtx, 2, bps)
+    # shape [batch_size, perm_num]
+    metric_of_ps = metric_of_ps_details.mean(dim=1)
+
+    # find the best metric and best permutation
+    best_metric, best_indexes = eval_func(metric_of_ps, dim=1)
+    best_indexes = best_indexes.detach()
+    best_perm = ps.T[best_indexes, :]
+    return best_metric, best_perm  # shape [batch], shape [batch, spk]
+
+
+def pit(
+    preds: torch.Tensor,
+    target: torch.Tensor,
+    metric_func: Callable,
+    eval_func: str = 'max',
+    **kwargs: Dict[str, Any]
+) -> Tuple[Tensor, Tensor]:
+    """ Permutation invariant training metric
+
+    Args:
+        target:
+            shape [batch, spk, ...]
+        preds:
+            shape [batch, spk, ...]
+        metric_func:
+            a metric function accept a batch of target and estimate,
+            i.e. metric_func(target[:, i, ...], estimate[:, j, ...]), and returns a batch of metric tensors [batch]
+        eval_func:
+            the function to find the best permutation, can be 'min' or 'max',
+            i.e. the smaller the better or the larger the better.
+        kwargs:
+            additional args for metric_func
+
+    Returns:
+        best_metric of shape [batch],
+        best_perm of shape [batch]
+
+    Example:
+        >>> import torch
+        >>> from torchmetrics.functional.audio import si_sdr, pit, permutate
+        >>> # [batch, spk, time]
+        >>> preds = torch.tensor([[[-0.0579,  0.3560, -0.9604], [-0.1719,  0.3205,  0.2951]]])
+        >>> target = torch.tensor([[[ 1.0958, -0.1648,  0.5228], [-0.4100,  1.1942, -0.5103]]])
+        >>> best_metric, best_perm = pit(preds, target, si_sdr, 'max')
+        >>> best_metric
+        tensor([-5.1091])
+        >>> best_perm
+        tensor([[0, 1]])
+        >>> preds_pmted = permutate(preds, best_perm)
+        >>> preds_pmted
+        tensor([[[-0.0579,  0.3560, -0.9604],
+                 [-0.1719,  0.3205,  0.2951]]])
+
+    Reference:
+        [1]	D. Yu, M. Kolbaek, Z.-H. Tan, J. Jensen, Permutation invariant training of deep models for
+        speaker-independent multi-talker speech separation, in: 2017 IEEE Int. Conf. Acoust. Speech
+        Signal Process. ICASSP, IEEE, New Orleans, LA, 2017: pp. 241–245. https://doi.org/10.1109/ICASSP.2017.7952154.
+    """
+    _check_same_shape(preds, target)
+    if eval_func not in ['max', 'min']:
+        raise RuntimeError('eval_func can only be "max" or "min"')
+    if len(target.shape) < 2:
+        raise RuntimeError(f"Inputs must be of shape [batch, spk, ...], got {target.shape} and {preds.shape} instead")
```

Comment:
remember to update tests!
```suggestion
        raise ValueError(f"Inputs must be of shape [batch, spk, ...], got {target.shape} and {preds.shape} instead")
```